src/interrupt/mod.rs

```diff
@@ -0,0 +1,210 @@
+///
+/// On x86 platforms, legacy interrupts means those interrupts routed through PICs or IOAPICs.
+#[cfg(feature = "legacy_irq")]
+#[derive(Copy, Clone, Debug)]
+pub struct LegacyIrqSourceConfig {}
+
+/// Configuration data for GenericMsi, PciMsi, PciMsix interrupts.
+#[cfg(feature = "msi_irq")]
+#[derive(Copy, Clone, Debug, Default)]
+pub struct MsiIrqSourceConfig {
+    /// High address to deliver message signaled interrupt.
+    pub high_addr: u32,
+    /// Low address to deliver message signaled interrupt.
+    pub low_addr: u32,
+    /// Data to write to deliver message signaled interrupt.
+    pub data: u32,
+}
+
+/// Trait to manage interrupt sources for virtual device backends.
+///
+/// The InterruptManager implementations should protect itself from concurrent accesses internally,
+/// so it could be invoked from multi-threaded context.
+pub trait InterruptManager {
+    /// Create an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object to manage
+    /// interrupt sources for a virtual device
+    ///
+    /// An [InterruptSourceGroup](trait.InterruptSourceGroup.html) object manages all interrupt
+    /// sources of the same type for a virtual device.
+    ///
+    /// # Arguments
+    /// * ty: type of interrupt source.
+    /// * base: base Interrupt Source ID to be managed by the group object.
+    /// * count: number of Interrupt Sources to be managed by the group object.
+    fn create_group(
+        &self,
+        ty: InterruptSourceType,
+        base: InterruptIndex,
+        count: InterruptIndex,
+    ) -> Result<Arc<Box<dyn InterruptSourceGroup>>>;
+
+    /// Destroy an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object created by
+    /// [create_group()](trait.InterruptManager.html#tymethod.create_group).
+    ///
+    /// Assume the caller takes the responsibility to disable all interrupt sources of the group
+    /// before calling destroy_group(). This assumption helps to simplify InterruptSourceGroup
+    /// implementations.
+    fn destroy_group(&self, group: Arc<Box<dyn InterruptSourceGroup>>) -> Result<()>;
+}
+
+/// Trait to manage a group of interrupt sources for a device.
+///
+/// A device may support several types of interrupts, and each type of interrupt may contain one or
+/// multiple continuous interrupt sources. For example, a PCI device may concurrently support:
+/// * Legacy Irq: exactly one interrupt source.
+/// * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+/// * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+///
+/// PCI MSI interrupts of a device may not be configured individually, and must configured as a
+/// whole block. So all interrupts of the same type of a device are abstracted as an
+/// [InterruptSourceGroup](trait.InterruptSourceGroup.html) object, instead of abstracting each
+/// interrupt source as a distinct InterruptSource.
+#[allow(clippy::len_without_is_empty)]
+#[allow(clippy::trivially_copy_pass_by_ref)]
+pub trait InterruptSourceGroup: Send + Sync {
```

Comment:
Not yet. I have had a quick glance at GICv3 spec, which look similar to x86 interrupt architecture.

Reply:
I have refined the implementation to better support arm/arm64, could you please help to take a look? @andreeaflorescu

Reply:
Can you point me to the changes? GitHub is not very good at showing diffs between versions.

Reply:
Basically I have moved legacy irq related code from src/interrupt/kvm/mod.rs into src/interrupt/kvm/legacy_irq.rs with better support of ARM/ARM64. For ARM/ARM64, the MSI part should be the same, and hopefully we only need to implement initialize_legacy().
```
    }

    #[cfg(any(target_arch = "aarch", target_arch = "aarch64"))]
    pub(super) fn initialize_legacy(
        _routes: &mut HashMap<u64, kvm_irq_routing_entry>,
    ) -> Result<()> {
        //TODO
        Ok(())
    }
```

Reply:
About that, I find it a bit too coupled, and I was currently working on a prototype on that. Can we split this PR in 2:
- one that defines just the traits and needed types
- one with implementation for Legacy and MSI?

The reason why I am asking this is because it seems that there's still missing functionality from the Legacy implementation. It would be nice to not merge uncomplete code. The other reason is that I think we should consider another way to specify the interrupt type. Having an enum defined in this crate doesn't allow for custom implementations on the interface. When we (Firecracker) will switch to the vm-device & vm-virtio crates, we might need a custom implementation for the InterruptSourceGroup. With the interface right now, that would not be possible because the interrupt type needs to be defined in vm-device.

Let's say that Firecracker wants to define the `InterruptSourceType` to `FirecrackerCustomInterrupt`. To do that, we would need to add `FirecrackerCustomInterrupt` to the enum which doesn't make sense. There are some other ways we can work around this problem, but all I can think of right now is to have InterruptSourceType defined as InterruptIndex (i.e. just a number). We can have some defines for common interrupt types. But anyway, we shouldn't need to think about this  just yet because having the interface merged should unblock us on vm-virtio for now.

What do you think?

Reply:
For the scenario mentioned above, how about 
1) disabling the feature kvm_irq.
2) providing firecracker specific implementation of InterruptManager and InterruptSourceGroup.

Actually we have two options here:
1) use feature flag to control the default kvm_irq implementation. If we want vmm specific implementation, we could disable feature.
2) create a new crate for the kvm_irq implementation. 

I don't like solution 2 too much because it introduces another crate and need to keep consistency with the vm-device crate.

Reply:
If you disable both features you get an empty InterruptSourceType because the interrupt type are configured out. Were you suggesting to raise an `unimplemented!` on the `interrupt_type()` method?

Also, since I noticed that the interrupt type enum variants are compiled out depending on the features, does that mean that in the virtio implementation you would depend on vm-device with all features enabled?

Reply:
Seems we are reaching the real interesting parts now:)
Essentially the PR defines two sets of features:
1) feature bit to control supported interrupt source types:
[features]
legacy_irq = []
msi_irq = []
pci_msi_irq = ["msi_irq"]
And there's one more coming feature bit of this class:
generic_msi_irq = ["msi_irq"]
So we could support three types of interrupt sources: legacy, pci msi and generic msi.
2) feature bit to control the implementations of InterruptManager/InterruptSourceGroup.
kvm_irq = ["kvm-ioctls", "kvm-bindings"]
The 'kvm_irq' enables a default implementation of InterruptManager/InterruptSourceGroup based KVM kernel module. If the vmm has special requirements, it may provide its own implementation by disabling the kvm_irq feature.

For the vm-virtio, there are three possible combinations:
1) use legacy irq only. The crosvm/firecracker project uses this mode.
2) use msi irq only. The cloud hypervisor works in this mode.
3) use both legacy irq and msi irq. Our dragonball vmm works in this mode, and we plan to switch to msi irq only mode.

Reply:
Please also refer to this kvm forum presentation for MMIO MSI support, https://kvmforum2019.sched.com/event/Tmw2/a-lightweight-virtual-interrupt-controller-for-containerserverless-jing-liu-chao-peng-intel?iframe=no&w=100%&sidebar=yes&bg=no

Reply:
Since this is a complex problem, I would really like to insist on splitting this PR. It looks like we have a few loose ends that we need to decide on. Let's first decide on the interface and have a separate PR only with that. Can you please add a PR only with the interface definition?

Reply:
I have different opinion here. 
A PR with both interface definition and concrete implementation may give more information about the design, and it's easier for community to do experiments with the design. 
If some design flaw has been discovered during experiments, we could easily change the interface design and implementation, or even revert the PR.
The vm-device is still in early stage, it would be better to share design and implementation sooner, otherwise we could not speed up the progress.

Reply:
I am not against prototyping. This is something that we discussed about a few times already.

The proposal that everyone seemed to be on board with was to have a PR that defines the interfaces and nothing more. Functionality was to be added in subsequent PRs.

The initial PR can include external links to prototypes with the interface being used in practice. Having such a large PR introduces a few problems:
1. Delays in reviewing and merging the code.
2. Key design points of the interface can be missed because we are not insisting on the interface.
3. Large PRs such as this one tend to reach around 100 comments. After a few iterations the comments get outdated, you cannot possibly go through all of them to see if they were fixed.